### ISSUE TITLE
ApiServer doesn't retry etcd connections in case of network partition

### ISSUE THREAD

**Issue by dhawal55** [kubernetes/kubernetes#59074]
/kind bug
/sig api-machinery

**What happened**: 
I'm running HA master with apiservers and etcd instances spread out on multi AZs. The apiserver have `--etcd-servers` option set to comma-separated list of all etcd servers. 
I isolated an AZ (using network ACLs) to simulate an AZ outage. The LoadBalancer in front of my apiserver removes the instance in the unreachable AZ. But some of the kubectl calls keep failing (timeout). It seems like apiserver is keeping long-lived connections to etcd and when the etcd instance is not reachable, it doesn't retry connecting to other etcd servers in the list.
 
**What you expected to happen**:
When an etcd instance goes down, apiserver should re-open connection to other healthy etcd servers in the list.

**How to reproduce it (as minimally and precisely as possible)**:
Ssh into one of the apiservers and run tcpdump and check which etcd instances it is talking too. Shut down those etcd instances. Run multiple kubectl calls and ones going to apiserver with connections to terminated etcd instances will timeout.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.7.6
- Cloud provider or hardware configuration: AWS
- OS (e.g. from /etc/os-release): Container Linux by CoreOS 1632.1.0 (Ladybug)
- Kernel (e.g. `uname -a`): Linux ip-x-x-x-x 4.14.11-coreos #1 SMP Fri Jan 5 08:05:10 UTC 2018 x86_64 Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz GenuineIntel GNU/Linux

**Comment by jennybuckley** [comment]
cc @mbohlool @jpbetz @wenjiaswe

**Comment by gtie** [comment]
We are getting bitten by this problem as well. The way I'm reproducing is:

1. Log on to an etcd node and kill its outbound networking (except for SSH)
```
[root@ip-10-121-10-134 ~]# iptables -I OUTPUT -p tcp --sport 22 -j ACCEPT
[root@ip-10-121-10-134 ~]# iptables -I OUTPUT 2 -j DROP;
[root@ip-10-121-10-134 ~]# date 
Do 8. Mär 11:13:17 UTC 2018
```
2. Verify the cluster componentstatus is timing out on one of the etcds:

```
$ kubectl get componentstatus
NAME                 STATUS      MESSAGE                                                                                                                                              ERROR
controller-manager   Healthy     ok                                                                                                                                                   
scheduler            Healthy     ok                                                                                                                                                   
etcd-2               Healthy     {"health": "true"}                                                                                                                                   
etcd-5               Healthy     {"health": "true"}                                                                                                                                   
etcd-0               Healthy     {"health": "true"}                                                                                                                                   
etcd-3               Healthy     {"health": "true"}                                                                                                                                   
etcd-1               Unhealthy   Get https://etcd-c.k8s.local:4001/health: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)   
etcd-4               Unhealthy   Get https://etcd-c.k8s.local:4002/health: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)   
```

2. Watch connections to etcd on the instances that run `apiserver`, e.g. quick and dirty `netstat -n | grep ':4001 ' | awk '{print $5 " - " $6}' | sort | uniq -c` long after the instance is dead:

```
Do 8. Mär 11:30:02 UTC 2018                                                                                                                                                                                    
     69 10.121.10.134:4001 - ESTABLISHED                                                                                                                                                                       
     69 10.121.8.139:4001 - ESTABLISHED                                                                                                                                                                        
     69 10.121.9.153:4001 - ESTABLISHED  
```

On our higher volume clusters, this situation leads to a buildup of "inflight requests" as reported by the Kubernetes API metrics, e.g.:
![image](https://user-images.githubusercontent.com/2247364/37153247-5b76cdfe-22dc-11e8-9f29-83c0b7246dc7.png)

The above happens on a Kubernetes 1.9.3 cluster with separate EC2 instances used for masters and for etcds (3 instances for each group).

The bottom line for us is that even though the etcd database survives fine with just 2/3 nodes up, we are getting Kubernetes API failures during that time.

**Comment by akumria** [comment]
I wonder if your problem goes away once https://github.com/kubernetes/kubernetes/pull/58008 is merged?

**Comment by johscheuer** [comment]
I assume you are running the apiserver with `--storage-backend=etcd3` ? Because we see the same behaviour with `etcd3` as storage-backend (I will test `etcd2` to validate if the MR from above solves this problem)

@akumria it seems like this commit will be in 1.10.*
